### PR TITLE
[autoparallel] adapt solver with self attention

### PR DESCRIPTION
**Pull request by YuliangLiu0306**
### What does this PR do

1. update the node merge rule.
2. update the CostGrpah class and Solver class to adapt the latest interface.
3. update the reshape utils to deal with the corner case.
4. upload the self-attention test file to reproduce the search result.

### CODE REVIEW

File: colossalai/auto_parallel/tensor_shard/solver/cost_graph.py [hunk 1]
```diff
@@ -63,14 +63,38 @@ def _build_cost_graph(self):
...
+            # parent_nodes = [node for node in strategies_vector.predecessor_nodes]
+            # children_nodes = [node for node in strategies_vector.successor_nodes]
+            parent_nodes = []
+            children_nodes = []
+
+            def _check_tensor_in_node(data):
+                """
+                This method is used to check whether the data has a tensor inside or not.
+                """
+                label = False
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/solver/cost_graph.py:75]: `label` -> `has_tensor_flag` for clarity.

File: colossalai/auto_parallel/tensor_shard/solver/cost_graph.py [hunk 2]
```diff
@@ -63,14 +63,38 @@ def _build_cost_graph(self):
...
+                if _check_tensor_in_node(node._meta_data):
+                    children_nodes.append(node)
+
             setattr(dst_node, 'parents', parent_nodes)
             setattr(dst_node, 'children', children_nodes)
 
             if self.simplify and strategies_vector.check_merge():
                 for followed_node in strategies_vector.predecessor_nodes:
-                    self.merge_pair.append((followed_node, dst_node))
+                    # we only merge node pairs which src node has a tensor element inside
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/solver/cost_graph.py:95]: Can you give a specific case to explain why this is necessary in the comment?
```
# we only merge node pairs which src node has a tensor element inside
# this is necessary because ...
# consider this case:
# ...
```

File: colossalai/auto_parallel/tensor_shard/solver/solver.py [hunk 3]
```diff
@@ -160,6 +160,7 @@ def _prepare_data_for_solver(self):
                     compute_cost = compute_cost_item.total
                     memory_cost = memory_cost_item.total
 
+                memory_cost = memory_cost.parameter + memory_cost.activation + memory_cost.buffer
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/solver/solver.py:163]: In line 157 and line 161, the variable `memory_cost` has been assigned, why do we overwrite it here?

File: colossalai/auto_parallel/tensor_shard/solver/cost_graph.py (hunk 1 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/solver/cost_graph.py:75]: Sure

File: colossalai/auto_parallel/tensor_shard/solver/cost_graph.py (hunk 2 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/solver/cost_graph.py:95]: OK

File: colossalai/auto_parallel/tensor_shard/solver/solver.py (hunk 3 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/solver/solver.py:163]: I will add some comments to explain it. In short, the memory_cost in 157 and 161 is in type of MemoryCost, and what solver wants is int, so I extract the related values in line 163.

Comment (FrankLeeeee) [reply colossalai/auto_parallel/tensor_shard/solver/solver.py:166]: Ok, I see. It is kind of dizzy to read without the comments haha

File: colossalai/auto_parallel/tensor_shard/solver/cost_graph.py (hunk 2 above)
Comment (FrankLeeeee) [reply colossalai/auto_parallel/tensor_shard/solver/cost_graph.py]: @YuliangLiu0306 this comment is not handled.

**Review by FrankLeeeee — APPROVED** [review]
